# Préparation des données

[Source originale du notebook tirée de *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio par Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Exploration des informations du `DataFrame`

> **Objectif d'apprentissage :** À la fin de cette sous-section, vous devriez être à l'aise pour trouver des informations générales sur les données stockées dans les DataFrames de pandas.

Une fois vos données chargées dans pandas, elles seront très probablement dans un `DataFrame`. Cependant, si le jeu de données dans votre `DataFrame` contient 60 000 lignes et 400 colonnes, comment commencer à comprendre ce avec quoi vous travaillez ? Heureusement, pandas offre des outils pratiques pour examiner rapidement des informations générales sur un `DataFrame`, ainsi que les premières et dernières lignes.

Pour explorer cette fonctionnalité, nous allons importer la bibliothèque Python scikit-learn et utiliser un jeu de données emblématique que tous les data scientists ont vu des centaines de fois : le jeu de données *Iris* du biologiste britannique Ronald Fisher, utilisé dans son article de 1936 "The use of multiple measurements in taxonomic problems" :


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
Nous avons chargé le jeu de données Iris dans la variable `iris_df`. Avant d'explorer les données, il serait utile de connaître le nombre de points de données que nous avons ainsi que la taille globale du jeu de données. Il est important de se faire une idée du volume de données avec lequel nous travaillons.


In [2]:
iris_df.shape

(150, 4)

Nous avons donc 150 lignes et 4 colonnes de données. Chaque ligne représente un point de données et chaque colonne représente une caractéristique unique associée au tableau de données. En gros, il y a 150 points de données contenant chacun 4 caractéristiques.

`shape` ici est un attribut du tableau de données et non une fonction, c'est pourquoi il ne se termine pas par une paire de parenthèses.


### `DataFrame.columns`
Passons maintenant aux 4 colonnes de données. Que représente exactement chacune d'elles ? L'attribut `columns` nous donnera le nom des colonnes dans le dataframe.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Comme nous pouvons le voir, il y a quatre (4) colonnes. L'attribut `columns` nous indique le nom des colonnes et rien d'autre. Cet attribut prend de l'importance lorsque nous voulons identifier les caractéristiques qu'un ensemble de données contient.


### `DataFrame.info`
La quantité de données (donnée par l'attribut `shape`) et le nom des caractéristiques ou colonnes (donné par l'attribut `columns`) nous donnent des informations sur le jeu de données. Maintenant, nous voudrions explorer le jeu de données plus en profondeur. La fonction `DataFrame.info()` est très utile pour cela.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


À partir d'ici, nous pouvons faire quelques observations :
1. Le type de données de chaque colonne : Dans ce jeu de données, toutes les données sont stockées sous forme de nombres à virgule flottante 64 bits.
2. Nombre de valeurs non nulles : La gestion des valeurs nulles est une étape importante dans la préparation des données. Cela sera traité plus tard dans le notebook.


### DataFrame.describe()
Supposons que nous ayons beaucoup de données numériques dans notre ensemble de données. Des calculs statistiques univariés tels que la moyenne, la médiane, les quartiles, etc., peuvent être effectués sur chacune des colonnes individuellement. La fonction `DataFrame.describe()` nous fournit un résumé statistique des colonnes numériques d'un ensemble de données.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

La sortie ci-dessus montre le nombre total de points de données, la moyenne, l'écart-type, le minimum, le quartile inférieur (25 %), la médiane (50 %), le quartile supérieur (75 %) et la valeur maximale de chaque colonne.


### `DataFrame.head`
Avec toutes les fonctions et attributs mentionnés ci-dessus, nous avons obtenu une vue d'ensemble du dataset. Nous savons combien de points de données il contient, combien de caractéristiques il possède, le type de données de chaque caractéristique et le nombre de valeurs non nulles pour chacune d'elles.

Il est maintenant temps de regarder les données elles-mêmes. Voyons à quoi ressemblent les premières lignes (les premiers points de données) de notre `DataFrame` :


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Comme le résultat ici, nous pouvons voir cinq (5) entrées du jeu de données. Si nous regardons l'index à gauche, nous découvrons que ce sont les cinq premières lignes.


### Exercice :

D'après l'exemple donné ci-dessus, il est évident que, par défaut, `DataFrame.head` retourne les cinq premières lignes d'un `DataFrame`. Dans la cellule de code ci-dessous, pouvez-vous trouver un moyen d'afficher plus de cinq lignes ?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Une autre façon d'examiner les données peut être à partir de la fin (au lieu du début). L'opposé de `DataFrame.head` est `DataFrame.tail`, qui renvoie les cinq dernières lignes d'un `DataFrame` :


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

En pratique, il est utile de pouvoir examiner facilement les premières lignes ou les dernières lignes d'un `DataFrame`, en particulier lorsque vous recherchez des valeurs aberrantes dans des ensembles de données ordonnés.

Toutes les fonctions et attributs présentés ci-dessus à l'aide d'exemples de code nous aident à avoir un aperçu et une idée du contenu des données.

> **À retenir :** Rien qu'en regardant les métadonnées sur les informations contenues dans un DataFrame ou les premières et dernières valeurs, vous pouvez avoir une idée immédiate de la taille, de la structure et du contenu des données avec lesquelles vous travaillez.


### Données manquantes
Plongeons dans le sujet des données manquantes. Les données manquantes surviennent lorsqu'aucune valeur n'est enregistrée dans certaines colonnes.

Prenons un exemple : disons qu'une personne est préoccupée par son poids et ne remplit pas le champ correspondant dans un sondage. Dans ce cas, la valeur du poids pour cette personne sera manquante.

La plupart du temps, dans les ensembles de données du monde réel, des valeurs manquantes apparaissent.

**Comment Pandas gère les données manquantes**

Pandas gère les valeurs manquantes de deux manières. La première, que vous avez déjà vue dans les sections précédentes, est `NaN`, ou Not a Number. Il s'agit en réalité d'une valeur spéciale faisant partie de la spécification IEEE des nombres à virgule flottante, et elle est uniquement utilisée pour indiquer des valeurs flottantes manquantes.

Pour les valeurs manquantes autres que les nombres flottants, pandas utilise l'objet Python `None`. Bien qu'il puisse sembler déroutant de rencontrer deux types de valeurs différentes qui expriment essentiellement la même chose, il existe des raisons programmatiques solides derrière ce choix de conception. En pratique, cette approche permet à pandas d'offrir un bon compromis dans la grande majorité des cas. Cela dit, `None` et `NaN` comportent des restrictions dont vous devez être conscient en ce qui concerne leur utilisation.


### `None`: données manquantes non flottantes
Étant donné que `None` provient de Python, il ne peut pas être utilisé dans les tableaux NumPy et pandas qui ne sont pas de type de données `'object'`. Rappelez-vous que les tableaux NumPy (et les structures de données dans pandas) ne peuvent contenir qu'un seul type de données. C'est ce qui leur confère une puissance considérable pour le traitement de données à grande échelle et les calculs, mais cela limite également leur flexibilité. Ces tableaux doivent être convertis au "plus petit dénominateur commun", c'est-à-dire au type de données qui englobe tout dans le tableau. Lorsque `None` est présent dans le tableau, cela signifie que vous travaillez avec des objets Python.

Pour voir cela en pratique, considérez l'exemple de tableau suivant (notez le `dtype` associé) :


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

La réalité des types de données surélevés entraîne deux conséquences. Premièrement, les opérations seront effectuées au niveau du code Python interprété plutôt que du code NumPy compilé. En d'autres termes, cela signifie que toutes les opérations impliquant des `Series` ou des `DataFrames` contenant `None` seront plus lentes. Bien que vous ne remarquiez probablement pas cet impact sur les performances, cela pourrait devenir problématique pour des ensembles de données volumineux.

La deuxième conséquence découle de la première. Étant donné que `None` ramène essentiellement les `Series` ou les `DataFrames` dans le monde du Python classique, l'utilisation d'agrégations NumPy/pandas comme `sum()` ou `min()` sur des tableaux contenant une valeur ``None`` produira généralement une erreur :


In [10]:
example1.sum()

TypeError: ignored

**Conclusion clé** : L'addition (et autres opérations) entre des entiers et des valeurs `None` est indéfinie, ce qui peut limiter ce que vous pouvez faire avec des ensembles de données qui en contiennent.


### `NaN` : valeurs flottantes manquantes

Contrairement à `None`, NumPy (et donc pandas) prend en charge `NaN` pour ses opérations vectorisées rapides et ses ufuncs. La mauvaise nouvelle est que toute opération arithmétique effectuée sur `NaN` donne toujours `NaN`. Par exemple :


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

La bonne nouvelle : les agrégations exécutées sur des tableaux contenant des `NaN` ne génèrent pas d'erreurs. La mauvaise nouvelle : les résultats ne sont pas uniformément utiles :


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Exercice :


In [11]:
# What happens if you add np.nan and None together?


Rappelez-vous : `NaN` est uniquement destiné aux valeurs flottantes manquantes ; il n'existe pas d'équivalent `NaN` pour les entiers, les chaînes ou les booléens.


### `NaN` et `None` : valeurs nulles dans pandas

Bien que `NaN` et `None` puissent se comporter légèrement différemment, pandas est conçu pour les gérer de manière interchangeable. Pour comprendre ce que nous voulons dire, considérez une `Series` d'entiers :


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Exercice :


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


Dans le processus de conversion des types de données pour établir une homogénéité des données dans les `Series` et les `DataFrame`s, pandas peut facilement alterner entre les valeurs manquantes `None` et `NaN`. En raison de cette caractéristique de conception, il peut être utile de considérer `None` et `NaN` comme deux variantes différentes de "null" dans pandas. En effet, certains des principaux outils que vous utiliserez pour gérer les valeurs manquantes dans pandas reflètent cette idée dans leurs noms :

- `isnull()`: Génère un masque booléen indiquant les valeurs manquantes
- `notnull()`: Opposé de `isnull()`
- `dropna()`: Renvoie une version filtrée des données
- `fillna()`: Renvoie une copie des données avec les valeurs manquantes remplies ou imputées

Ce sont des méthodes importantes à maîtriser et à utiliser avec aisance, alors examinons chacune d'elles en détail.


### Détection des valeurs nulles

Maintenant que nous avons compris l'importance des valeurs manquantes, nous devons les détecter dans notre ensemble de données avant de les traiter.  
Les méthodes `isnull()` et `notnull()` sont vos principales options pour détecter les données nulles. Elles renvoient toutes deux des masques booléens sur vos données.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Regardez attentivement le résultat. Est-ce que quelque chose vous surprend ? Bien que `0` soit un nul arithmétique, c'est néanmoins un entier parfaitement valide, et pandas le traite comme tel. `''` est un peu plus subtil. Bien que nous l'ayons utilisé dans la Section 1 pour représenter une valeur de chaîne vide, c'est néanmoins un objet de type chaîne et non une représentation de nul selon pandas.

Maintenant, retournons la situation et utilisons ces méthodes d'une manière plus proche de leur utilisation pratique. Vous pouvez utiliser des masques booléens directement comme index de ``Series`` ou ``DataFrame``, ce qui peut être utile lorsque vous essayez de travailler avec des valeurs manquantes (ou présentes) isolées.

Si nous voulons le nombre total de valeurs manquantes, nous pouvons simplement effectuer une somme sur le masque produit par la méthode `isnull()`.


In [19]:
example3.isnull().sum()

2

### Exercice :


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Conclusion clé** : Les méthodes `isnull()` et `notnull()` produisent des résultats similaires lorsque vous les utilisez dans des DataFrames : elles affichent les résultats ainsi que l'index de ces résultats, ce qui vous sera d'une grande aide lorsque vous travaillez avec vos données.


### Gérer les données manquantes

> **Objectif d'apprentissage :** À la fin de cette sous-section, vous devriez savoir comment et quand remplacer ou supprimer les valeurs nulles dans les DataFrames.

Les modèles de Machine Learning ne peuvent pas gérer eux-mêmes les données manquantes. Ainsi, avant de transmettre les données au modèle, nous devons traiter ces valeurs manquantes.

La manière dont les données manquantes sont traitées implique des compromis subtils, pouvant influencer votre analyse finale et les résultats dans le monde réel.

Il existe principalement deux façons de gérer les données manquantes :

1.   Supprimer la ligne contenant la valeur manquante
2.   Remplacer la valeur manquante par une autre valeur

Nous discuterons de ces deux méthodes ainsi que de leurs avantages et inconvénients en détail.


### Suppression des valeurs nulles

La quantité de données que nous transmettons à notre modèle a un effet direct sur ses performances. Supprimer les valeurs nulles signifie que nous réduisons le nombre de points de données, et donc la taille du jeu de données. Il est donc conseillé de supprimer les lignes contenant des valeurs nulles lorsque le jeu de données est assez grand.

Un autre cas pourrait être qu'une certaine ligne ou colonne contient beaucoup de valeurs manquantes. Dans ce cas, elles peuvent être supprimées car elles n'apporteraient pas beaucoup de valeur à notre analyse, étant donné que la plupart des données sont manquantes pour cette ligne/colonne.

Au-delà de l'identification des valeurs manquantes, pandas offre un moyen pratique de supprimer les valeurs nulles des `Series` et des `DataFrame`. Pour voir cela en pratique, revenons à `example3`. La fonction `DataFrame.dropna()` permet de supprimer les lignes contenant des valeurs nulles.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Notez que cela devrait ressembler à votre sortie de `example3[example3.notnull()]`. La différence ici est que, plutôt que de simplement indexer les valeurs masquées, `dropna` a supprimé ces valeurs manquantes de la `Series` `example3`.

Étant donné que les DataFrames ont deux dimensions, ils offrent plus d'options pour supprimer des données.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Avez-vous remarqué que pandas a converti deux des colonnes en flottants pour gérer les `NaN` ?)

Vous ne pouvez pas supprimer une seule valeur d'un `DataFrame`, vous devez donc supprimer des lignes ou des colonnes entières. Selon ce que vous faites, vous pourriez préférer l'une ou l'autre option, et pandas vous offre des possibilités pour les deux. Étant donné qu'en science des données, les colonnes représentent généralement des variables et les lignes des observations, il est plus courant de supprimer des lignes de données ; le paramètre par défaut de `dropna()` est de supprimer toutes les lignes contenant des valeurs nulles :


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Si nécessaire, vous pouvez supprimer les valeurs NA des colonnes. Utilisez `axis=1` pour ce faire :


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Notez que cela peut supprimer beaucoup de données que vous pourriez vouloir conserver, en particulier dans des ensembles de données plus petits. Que faire si vous souhaitez simplement supprimer les lignes ou colonnes contenant plusieurs valeurs nulles, voire toutes ? Vous pouvez spécifier ces paramètres dans `dropna` avec les arguments `how` et `thresh`.

Par défaut, `how='any'` (si vous souhaitez vérifier par vous-même ou voir quels autres paramètres la méthode propose, exécutez `example4.dropna?` dans une cellule de code). Vous pouvez également spécifier `how='all'` pour ne supprimer que les lignes ou colonnes contenant uniquement des valeurs nulles. Étendons notre exemple de `DataFrame` pour voir cela en action dans le prochain exercice.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Points clés :  
1. Supprimer les valeurs nulles est une bonne idée uniquement si le jeu de données est suffisamment grand.  
2. Des lignes ou colonnes entières peuvent être supprimées si la majorité de leurs données sont manquantes.  
3. La méthode `DataFrame.dropna(axis=)` permet de supprimer les valeurs nulles. L'argument `axis` indique si ce sont les lignes ou les colonnes qui doivent être supprimées.  
4. L'argument `how` peut également être utilisé. Par défaut, il est défini sur `any`. Ainsi, seules les lignes/colonnes contenant des valeurs nulles sont supprimées. Il peut être défini sur `all` pour spécifier que seules les lignes/colonnes où toutes les valeurs sont nulles seront supprimées.  


### Exercice :


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


Le paramètre `thresh` vous offre un contrôle plus précis : vous définissez le nombre de valeurs *non nulles* qu'une ligne ou une colonne doit avoir pour être conservée :


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

Ici, la première et la dernière ligne ont été supprimées, car elles contiennent seulement deux valeurs non nulles.


### Remplir les valeurs nulles

Il est parfois logique de remplacer les valeurs manquantes par des valeurs qui pourraient être valides. Il existe plusieurs techniques pour remplir les valeurs nulles. La première consiste à utiliser les connaissances du domaine (connaissance du sujet sur lequel repose le jeu de données) pour estimer d'une manière ou d'une autre les valeurs manquantes.

Vous pourriez utiliser `isnull` pour effectuer cette opération directement, mais cela peut être fastidieux, surtout si vous avez beaucoup de valeurs à remplir. Étant donné que cette tâche est très courante en science des données, pandas propose `fillna`, qui retourne une copie de la `Series` ou du `DataFrame` avec les valeurs manquantes remplacées par celles de votre choix. Créons un autre exemple de `Series` pour voir comment cela fonctionne en pratique.


### Données catégoriques (non numériques)
Commençons par les données non numériques. Dans les ensembles de données, nous avons des colonnes contenant des données catégoriques. Par exemple, le genre, Vrai ou Faux, etc.

Dans la plupart de ces cas, nous remplaçons les valeurs manquantes par la `mode` de la colonne. Supposons que nous avons 100 points de données, dont 90 indiquent Vrai, 8 indiquent Faux et 2 n'ont pas répondu. Alors, nous pouvons remplir les 2 valeurs manquantes avec Vrai, en prenant en compte l'ensemble de la colonne.

Encore une fois, ici, nous pouvons utiliser des connaissances spécifiques au domaine. Prenons un exemple de remplissage avec la mode.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Maintenant, trouvons d'abord la mode avant de remplir la valeur `None` avec la mode.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

Donc, nous remplacerons None par True


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Comme nous pouvons le voir, la valeur nulle a été remplacée. Inutile de dire que nous aurions pu écrire n'importe quoi à la place de `'True'` et cela aurait été substitué.


### Données numériques
Passons maintenant aux données numériques. Ici, nous avons deux méthodes courantes pour remplacer les valeurs manquantes :

1. Remplacer par la médiane de la ligne  
2. Remplacer par la moyenne de la ligne  

On remplace par la médiane dans le cas de données asymétriques avec des valeurs aberrantes. En effet, la médiane est robuste face aux valeurs aberrantes.

Lorsque les données sont normalisées, on peut utiliser la moyenne, car dans ce cas, la moyenne et la médiane seront assez proches.

Prenons d'abord une colonne qui suit une distribution normale et remplissons les valeurs manquantes avec la moyenne de la colonne.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

La moyenne de la colonne est


In [33]:
np.mean(fill_with_mean[0])

0.0

Remplissage avec la moyenne


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Comme nous pouvons le voir, la valeur manquante a été remplacée par sa moyenne.


Maintenant, essayons un autre dataframe, et cette fois nous remplacerons les valeurs None par la médiane de la colonne.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

La médiane de la deuxième colonne est


In [36]:
fill_with_median[1].median()

4.0

Remplissage avec la médiane


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Comme nous pouvons le voir, la valeur NaN a été remplacée par la médiane de la colonne.


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Vous pouvez remplir toutes les entrées nulles avec une seule valeur, comme `0` :


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Points clés :
1. Remplir les valeurs manquantes doit être envisagé lorsqu'il y a peu de données ou lorsqu'une stratégie est définie pour combler ces lacunes.
2. Les connaissances du domaine peuvent être utilisées pour estimer et compléter les valeurs manquantes.
3. Pour les données catégoriques, les valeurs manquantes sont généralement remplacées par la mode de la colonne.
4. Pour les données numériques, les valeurs manquantes sont souvent remplacées par la moyenne (pour les ensembles de données normalisés) ou la médiane des colonnes.


### Exercice :


In [40]:
# What happens if you try to fill null values with a string, like ''?


Vous pouvez **remplir par propagation** les valeurs nulles, ce qui consiste à utiliser la dernière valeur valide pour remplir une valeur nulle :


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Vous pouvez également **remplir en arrière** pour propager la prochaine valeur valide en arrière afin de remplir une valeur nulle :


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Comme vous pouvez le deviner, cela fonctionne de la même manière avec les DataFrames, mais vous pouvez également spécifier un `axis` le long duquel remplir les valeurs nulles :


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Remarquez que lorsqu'une valeur précédente n'est pas disponible pour le remplissage vers l'avant, la valeur nulle reste.


### Exercice :


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Vous pouvez être créatif quant à la manière d'utiliser `fillna`. Par exemple, examinons à nouveau `example4`, mais cette fois remplissons les valeurs manquantes avec la moyenne de toutes les valeurs dans le `DataFrame` :


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Remarquez que la colonne 3 est toujours sans valeur : la direction par défaut est de remplir les valeurs ligne par ligne.

> **À retenir :** Il existe plusieurs façons de gérer les valeurs manquantes dans vos ensembles de données. La stratégie spécifique que vous utilisez (les supprimer, les remplacer, ou même la manière dont vous les remplacez) doit être dictée par les particularités de ces données. Vous développerez une meilleure intuition pour traiter les valeurs manquantes à mesure que vous manipulerez et interagirez avec des ensembles de données.


### Encodage des données catégorielles

Les modèles d'apprentissage automatique ne traitent que des chiffres et de toute forme de données numériques. Ils ne peuvent pas faire la différence entre un Oui et un Non, mais ils peuvent distinguer entre 0 et 1. Ainsi, après avoir rempli les valeurs manquantes, nous devons encoder les données catégorielles sous une forme numérique pour que le modèle puisse les comprendre.

L'encodage peut être réalisé de deux manières. Nous allons les examiner ensuite.


**ENCODAGE DES ÉTIQUETTES**

L'encodage des étiquettes consiste essentiellement à convertir chaque catégorie en un numéro. Par exemple, supposons que nous avons un ensemble de données sur les passagers d'une compagnie aérienne et qu'il y a une colonne indiquant leur classe parmi les suivantes ['classe affaires', 'classe économique', 'première classe']. Si un encodage des étiquettes est effectué, cela serait transformé en [0,1,2]. Voyons un exemple avec du code. Comme nous allons apprendre `scikit-learn` dans les prochains notebooks, nous ne l'utiliserons pas ici.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Pour effectuer l'encodage des étiquettes sur la première colonne, nous devons d'abord décrire une correspondance entre chaque classe et un numéro, avant de remplacer.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Comme nous pouvons le constater, le résultat correspond à ce que nous avions prévu. Alors, quand utilise-t-on l'encodage des étiquettes ? L'encodage des étiquettes est utilisé dans l'un ou les deux cas suivants :
1. Lorsque le nombre de catégories est élevé
2. Lorsque les catégories sont ordonnées.


**ENCODAGE ONE HOT**

Un autre type d'encodage est l'encodage One Hot. Dans ce type d'encodage, chaque catégorie de la colonne est ajoutée comme une colonne distincte, et chaque point de données recevra un 0 ou un 1 en fonction de la présence ou non de cette catégorie. Ainsi, s'il y a n catégories différentes, n colonnes seront ajoutées au dataframe.

Par exemple, prenons le même exemple des classes d'avion. Les catégories étaient : ['classe affaires', 'classe économique', 'première classe']. Donc, si nous effectuons un encodage One Hot, les trois colonnes suivantes seront ajoutées au jeu de données : ['classe_classe affaires', 'classe_classe économique', 'classe_première classe'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Effectuons un encodage one hot sur la première colonne


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Chaque colonne encodée en one-hot contient 0 ou 1, ce qui indique si cette catégorie existe pour ce point de données.


Quand utilise-t-on l'encodage one hot ? L'encodage one hot est utilisé dans l'un ou les deux cas suivants :

1. Lorsque le nombre de catégories et la taille du jeu de données sont réduits.
2. Lorsque les catégories ne suivent aucun ordre particulier.


> Points clés :
1. Le codage est effectué pour convertir des données non numériques en données numériques.
2. Il existe deux types de codage : le codage par étiquettes et le codage One Hot, qui peuvent être réalisés en fonction des besoins du jeu de données.


## Suppression des données dupliquées

> **Objectif d'apprentissage :** À la fin de cette sous-section, vous devriez être à l'aise pour identifier et supprimer les valeurs dupliquées dans les DataFrames.

En plus des données manquantes, vous rencontrerez souvent des données dupliquées dans les ensembles de données réels. Heureusement, pandas offre un moyen simple de détecter et de supprimer les entrées dupliquées.


### Identifier les doublons : `duplicated`

Vous pouvez facilement repérer les valeurs en double en utilisant la méthode `duplicated` de pandas, qui renvoie un masque booléen indiquant si une entrée dans un `DataFrame` est un doublon d'une précédente. Créons un autre exemple de `DataFrame` pour voir cela en pratique.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Suppression des doublons : `drop_duplicates`
`drop_duplicates` renvoie simplement une copie des données pour lesquelles toutes les valeurs `duplicated` sont `False` :


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Les fonctions `duplicated` et `drop_duplicates` considèrent par défaut toutes les colonnes, mais vous pouvez spécifier qu'elles examinent uniquement un sous-ensemble de colonnes dans votre `DataFrame` :


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **Conclusion :** Supprimer les données dupliquées est une étape essentielle de presque tous les projets de science des données. Les données dupliquées peuvent modifier les résultats de vos analyses et vous fournir des résultats inexacts !


## Vérifications de la qualité des données dans le monde réel

> **Objectif d'apprentissage :** À la fin de cette section, vous devriez être à l'aise pour détecter et corriger les problèmes courants de qualité des données dans le monde réel, notamment les valeurs catégoriques incohérentes, les valeurs numériques anormales (valeurs aberrantes) et les entités dupliquées avec des variations.

Bien que les valeurs manquantes et les doublons exacts soient des problèmes fréquents, les ensembles de données réels contiennent souvent des problèmes plus subtils :

1. **Valeurs catégoriques incohérentes** : Une même catégorie écrite différemment (par exemple, "USA", "U.S.A", "United States")
2. **Valeurs numériques anormales** : Des valeurs extrêmes qui indiquent des erreurs de saisie (par exemple, âge = 999)
3. **Lignes quasi-duplicatées** : Des enregistrements représentant la même entité avec de légères variations

Explorons les techniques pour détecter et gérer ces problèmes.


### Création d'un exemple de jeu de données "sale"

Tout d'abord, créons un exemple de jeu de données contenant les types de problèmes que l'on rencontre fréquemment dans les données du monde réel :


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. Détection des valeurs catégoriques incohérentes

Remarquez que la colonne `country` contient plusieurs représentations pour les mêmes pays. Identifions ces incohérences :


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### Standardisation des valeurs catégoriques

Nous pouvons créer une correspondance pour standardiser ces valeurs. Une approche simple consiste à convertir en minuscules et à créer un dictionnaire de correspondance :


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**Alternative : Utiliser la correspondance approximative**

Pour des cas plus complexes, nous pouvons utiliser la correspondance de chaînes approximative avec la bibliothèque `rapidfuzz` pour détecter automatiquement des chaînes similaires :


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. Détection des valeurs numériques anormales (Outliers)

En examinant la colonne `age`, nous avons des valeurs suspectes comme 199 et -5. Utilisons des méthodes statistiques pour détecter ces valeurs aberrantes.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### Utilisation de la méthode IQR (Intervalle Interquartile)

La méthode IQR est une technique statistique robuste pour la détection des valeurs aberrantes, moins sensible aux valeurs extrêmes :


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Utilisation de la méthode du score Z

La méthode du score Z identifie les valeurs aberrantes en fonction des écarts-types par rapport à la moyenne :


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### Gestion des valeurs aberrantes

Une fois détectées, les valeurs aberrantes peuvent être traitées de plusieurs manières :
1. **Supprimer** : Retirer les lignes contenant des valeurs aberrantes (si ce sont des erreurs)
2. **Limiter** : Remplacer par des valeurs limites
3. **Remplacer par NaN** : Considérer comme des données manquantes et utiliser des techniques d'imputation
4. **Conserver** : Si ce sont des valeurs extrêmes légitimes


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. Détection des lignes quasi-duplicatées

Remarquez que notre ensemble de données contient plusieurs entrées pour "John Smith" avec des valeurs légèrement différentes. Identifions les doublons potentiels en nous basant sur la similarité des noms.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### Recherche de quasi-doublons avec la correspondance approximative

Pour une détection de doublons plus avancée, nous pouvons utiliser la correspondance approximative pour trouver des noms similaires :


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### Gestion des doublons

Une fois identifiés, vous devez décider comment gérer les doublons :
1. **Conserver la première occurrence** : Utilisez `drop_duplicates(keep='first')`
2. **Conserver la dernière occurrence** : Utilisez `drop_duplicates(keep='last')`
3. **Agréger les informations** : Combiner les informations des lignes en doublon
4. **Revue manuelle** : Marquer pour une vérification humaine


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### Résumé : Pipeline complet de nettoyage des données

Regroupons tout cela dans un pipeline de nettoyage complet :


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 Exercice de défi

C'est à vous de jouer ! Voici une nouvelle ligne de données avec plusieurs problèmes de qualité. Pouvez-vous :

1. Identifier tous les problèmes dans cette ligne
2. Écrire du code pour corriger chaque problème
3. Ajouter la ligne nettoyée au jeu de données

Voici les données problématiques :


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### Points Clés

1. **Catégories incohérentes** sont fréquentes dans les données réelles. Vérifiez toujours les valeurs uniques et standardisez-les à l'aide de correspondances ou de méthodes de rapprochement approximatif.

2. **Les valeurs aberrantes** peuvent avoir un impact significatif sur votre analyse. Utilisez vos connaissances du domaine combinées à des méthodes statistiques (IQR, score Z) pour les détecter.

3. **Les quasi-duplicatas** sont plus difficiles à repérer que les duplicatas exacts. Pensez à utiliser des techniques de rapprochement approximatif et à normaliser les données (mise en minuscule, suppression des espaces) pour les identifier.

4. **Le nettoyage des données est itératif**. Vous devrez peut-être appliquer plusieurs techniques et examiner les résultats avant de finaliser votre ensemble de données nettoyé.

5. **Documentez vos décisions**. Notez les étapes de nettoyage que vous avez appliquées et les raisons, car cela est essentiel pour la reproductibilité et la transparence.

> **Bonne pratique :** Conservez toujours une copie de vos données "brutes" originales. Ne remplacez jamais vos fichiers sources - créez des versions nettoyées avec des conventions de nommage claires comme `data_cleaned.csv`.



---

**Avertissement** :  
Ce document a été traduit à l'aide du service de traduction automatique [Co-op Translator](https://github.com/Azure/co-op-translator). Bien que nous nous efforcions d'assurer l'exactitude, veuillez noter que les traductions automatisées peuvent contenir des erreurs ou des inexactitudes. Le document original dans sa langue d'origine doit être considéré comme la source faisant autorité. Pour des informations critiques, il est recommandé de recourir à une traduction humaine professionnelle. Nous ne sommes pas responsables des malentendus ou des interprétations erronées résultant de l'utilisation de cette traduction.
